### ISSUE TITLE
Training question when change the backbone

### ISSUE THREAD

**Issue by connernam** [Daniil-Osokin/lightweight-human-pose-estimation.pytorch#235]
Hi, I have a few questions,
I want to change the backbone to MobilenetV3.
However, there was a problem in implementing it by referring to the training code of the repo.

First, the loss comes out too high.
Second, after validation, both AP and AR are output at 0.000.

I downloaded the pretrained model of pytorch at https://download.pytorch.org/models/mobilenet_v3_small-047dcff4.pth
Then, all weights except classifier were grouped into 'state_dict' to create a new pth file.

And I used the code of https://github.com/pytorch/vision/blob/11bf27e37190b320216c349e39b085fb33aefed1/torchvision/models/mobilenetv3.py.

And I revised the code as below
![image](https://user-images.githubusercontent.com/42915968/150474282-e2222987-fde1-4e86-886a-bfc33fcf5a52.png)
![image](https://user-images.githubusercontent.com/42915968/150474326-1796c871-b73c-4e84-8c78-fddc5d3c47d4.png)
![image](https://user-images.githubusercontent.com/42915968/150474347-d578ec47-f14f-4aab-bb5f-37205e638b34.png)
![image](https://user-images.githubusercontent.com/42915968/150474600-6b1ae07e-3cf6-473a-823f-9c34176ab989.png)

And I changed the stride from 8 to 16 in the train code and the Val code.

After that, when the learning proceeded, the following results were met.
![image](https://user-images.githubusercontent.com/42915968/150474919-b5c9ca6a-5619-4ebe-928d-297fc1c232d6.png)

I proceeded the same as Training steps 1, 2, and 3 in readme, and detection.json was also generated normally in val code.

I don't know where the problem is, but can I tell you what to check?

**Comment by connernam** [comment]
Add my train.py code

import argparse
import cv2
import os

import torch
from torch.nn import DataParallel
import torch.optim as optim
from torch.utils.data import DataLoader
from torchvision import transforms

from datasets.coco import CocoTrainDataset
from datasets.transformations import ConvertKeypoints, Scale, Rotate, CropPad, Flip
from modules.get_parameters import get_parameters_conv, get_parameters_bn, get_parameters_conv_depthwise
# from models.with_mobilenet import PoseEstimationWithMobileNet
from models.with_mobileNetV3 import mobilenet_v3_small
from modules.loss import l2_loss
from modules.load_state import load_state, load_from_mobilenet
from val import evaluate

cv2.setNumThreads(0)
cv2.ocl.setUseOpenCL(False)  # To prevent freeze of DataLoader


def train(prepared_train_labels, train_images_folder, num_refinement_stages, base_lr, batch_size, batches_per_iter,
          num_workers, checkpoint_path, weights_only, from_mobilenet, checkpoints_folder, log_after,
          val_labels, val_images_folder, val_output_name, checkpoint_after, val_after):

    net = mobilenet_v3_small()
    stride = 16  # 8 -> 32
    sigma = 7
    path_thickness = 1

    dataset = CocoTrainDataset(prepared_train_labels,
                               train_images_folder,
                               stride, sigma, path_thickness,
                               transform=transforms.Compose([
                                   ConvertKeypoints(),
                                   Scale(),
                                   Rotate(pad=(128, 128, 128)),
                                   CropPad(pad=(128, 128, 128)),
                                   Flip()]))
    train_loader = DataLoader(dataset, batch_size=batch_size, shuffle=True, num_workers=num_workers)

    optimizer = optim.Adam([
        {'params': get_parameters_conv(net.features, 'weight')},
        {'params': get_parameters_conv(net.features, 'bias'), 'lr': base_lr * 2, 'weight_decay': 0},
        {'params': get_parameters_conv(net.cpm, 'weight'), 'lr': base_lr},
        {'params': get_parameters_conv(net.cpm, 'bias'), 'lr': base_lr * 2, 'weight_decay': 0},
        {'params': get_parameters_conv_depthwise(net.cpm, 'weight'), 'weight_decay': 0},
        {'params': get_parameters_conv(net.initial_stage, 'weight'), 'lr': base_lr},
        {'params': get_parameters_conv(net.initial_stage, 'bias'), 'lr': base_lr * 2, 'weight_decay': 0},
        {'params': get_parameters_conv(net.refinement_stages, 'weight'), 'lr': base_lr * 4},
        {'params': get_parameters_conv(net.refinement_stages, 'bias'), 'lr': base_lr * 8, 'weight_decay': 0},
        {'params': get_parameters_bn(net.refinement_stages, 'weight'), 'weight_decay': 0},
        {'params': get_parameters_bn(net.refinement_stages, 'bias'), 'lr': base_lr * 2, 'weight_decay': 0},
        ], lr=base_lr, weight_decay=5e-4)

    num_iter = 0
    current_epoch = 0
    drop_after_epoch = [100, 200, 260]
    scheduler = optim.lr_scheduler.MultiStepLR(optimizer, milestones=drop_after_epoch, gamma=0.333)
    if checkpoint_path:
        checkpoint = torch.load(checkpoint_path)

        if from_mobilenet:
            load_from_mobilenet(net, checkpoint)
        else:
            load_state(net, checkpoint)
            if not weights_only:
                optimizer.load_state_dict(checkpoint['optimizer'])
                scheduler.load_state_dict(checkpoint['scheduler'])
                num_iter = checkpoint['iter']
                current_epoch = checkpoint['current_epoch']

    print('start')

    net = DataParallel(net).cuda()
    net.train()

    for epochId in range(current_epoch, 100):
        # scheduler.step()
        optimizer.step()
        total_losses = [0, 0] * (num_refinement_stages + 1)  # heatmaps loss, paf loss per stage
        batch_per_iter_idx = 0
        for batch_data in train_loader:
            if batch_per_iter_idx == 0:
                optimizer.zero_grad()

            images = batch_data['image'].cuda()
            keypoint_masks = batch_data['keypoint_mask'].cuda()
            paf_masks = batch_data['paf_mask'].cuda()
            keypoint_maps = batch_data['keypoint_maps'].cuda()
            paf_maps = batch_data['paf_maps'].cuda()

            stages_output = net(images)
            losses = []

            for loss_idx in range(len(total_losses) // 2):
                losses.append(l2_loss(stages_output[loss_idx * 2], keypoint_maps, keypoint_masks, images.shape[0]))
                losses.append(l2_loss(stages_output[loss_idx * 2 + 1], paf_maps, paf_masks, images.shape[0]))
                total_losses[loss_idx * 2] += losses[-2].item() / batches_per_iter
                total_losses[loss_idx * 2 + 1] += losses[-1].item() / batches_per_iter

            loss = losses[0]
            for loss_idx in range(1, len(losses)):
                loss += losses[loss_idx]
            loss /= batches_per_iter
            loss.backward()
            batch_per_iter_idx += 1
            if batch_per_iter_idx == batches_per_iter:
                # optimizer.step()
                scheduler.step()
                batch_per_iter_idx = 0
                num_iter += 1
            else:
                continue

            if num_iter % log_after == 0:
                print('Iter: {}'.format(num_iter))
                for loss_idx in range(len(total_losses) // 2):
                    print('\n'.join(['stage{}_pafs_loss:     {}', 'stage{}_heatmaps_loss: {}']).format(
                        loss_idx + 1, total_losses[loss_idx * 2 + 1] / log_after,
                        loss_idx + 1, total_losses[loss_idx * 2] / log_after))
                for loss_idx in range(len(total_losses)):
                    total_losses[loss_idx] = 0
            if num_iter % checkpoint_after == 0:
                snapshot_name = '{}/checkpoint_iter_{}.pth'.format(checkpoints_folder, num_iter)
                torch.save({'state_dict': net.module.state_dict(),
                            'optimizer': optimizer.state_dict(),
                            'scheduler': scheduler.state_dict(),
                            'iter': num_iter,
                            'current_epoch': epochId},
                           snapshot_name)
            if num_iter % val_after == 0:
                print('Validation...')
                evaluate(val_labels, val_output_name, val_images_folder, net)
                net.train()

# ------------------------------------------------------------------------------------------------------------


if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--prepared-train-labels', type=str, default='./prepared_train_annotation.pkl',
                        help='path to the file with prepared annotations')
    parser.add_argument('--train-images-folder', type=str, default='F:/dataset/COCO/train2017', help='path to COCO train images folder')
    parser.add_argument('--num-refinement-stages', type=int, default=1, help='number of refinement stages')
    parser.add_argument('--base-lr', type=float, default=4e-5, help='initial learning rate')
    parser.add_argument('--batch-size', type=int, default=20, help='batch size')
    parser.add_argument('--batches-per-iter', type=int, default=4, help='number of batches to accumulate gradient from')
    parser.add_argument('--num-workers', type=int, default=8, help='number of workers')
    parser.add_argument('--checkpoint-path', type=str, required=True, help='path to the checkpoint to continue training from')
    parser.add_argument('--from-mobilenet', action='store_true',
                        help='load weights from mobilenet feature extractor')
    parser.add_argument('--weights-only', action='store_true',
                        help='just initialize layers with pre-trained weights and start training from the beginning')
    parser.add_argument('--experiment-name', type=str, default='default',
                        help='experiment name to create folder for checkpoints')
    parser.add_argument('--log-after', type=int, default=100, help='number of iterations to print train loss')

    parser.add_argument('--val-labels', type=str, default='./val_subset.json', help='path to json with keypoints val labels')
    parser.add_argument('--val-images-folder', type=str, default='F:/dataset/COCO/val2017', help='path to COCO val images folder')
    parser.add_argument('--val-output-name', type=str, default='detections.json',
                        help='name of output json file with detected keypoints')
    parser.add_argument('--checkpoint-after', type=int, default=1000,
                        help='number of iterations to save checkpoint')
    parser.add_argument('--val-after', type=int, default=100,
                        help='number of iterations to run validation')
    args = parser.parse_args()

    checkpoints_folder = '{}_checkpoints'.format(args.experiment_name)
    if not os.path.exists(checkpoints_folder):
        os.makedirs(checkpoints_folder)

    train(args.prepared_train_labels, args.train_images_folder, args.num_refinement_stages, args.base_lr, args.batch_size,
          args.batches_per_iter, args.num_workers, args.checkpoint_path, args.weights_only, args.from_mobilenet,
          checkpoints_folder, args.log_after, args.val_labels, args.val_images_folder, args.val_output_name,
          args.checkpoint_after, args.val_after)

**Comment by Daniil-Osokin** [comment]
Hi! Modifications sound reasonable, it is hard to tell what went in a wrong way without debugging. My guess is to check generation of the target heatmaps first. Just visualize them alongside with the image to see if keypoints and pafs are on corresponding places. Changing stride might affect generation of the target data.

**Comment by alicera** [comment]
target heatmaps first.
Do you have the reference?

**Comment by Daniil-Osokin** [comment]
The reference is produced by the training code.

**Comment by Daniil-Osokin** [comment]
Hope, it helped.

**Comment by connernam** [comment]
Sorry, I late responce.
I tried outputting the target's heatmap as shown in the code below.

![image](https://user-images.githubusercontent.com/42915968/152900193-530a99b8-154a-45b0-828d-305bb43a61e1.png)

The results are as follows.

![image](https://user-images.githubusercontent.com/42915968/152900297-50403abe-6cdb-4c44-9cf2-72cd97d46010.png)

I think it's being printed normally. Is that right?

**Comment by connernam** [comment]
The generation of the target heatmap seems normal. If so, the loss of the input heatmap seems to be an abnormal cause.
I tried outputting the input heatmap.

![image](https://user-images.githubusercontent.com/42915968/153553355-67cc8274-2a79-42e0-8a3f-510545603b6f.png)

There seems to be something wrong with the input heatmap.

I would appreciate it if you could let me know which part to check.

**Comment by Daniil-Osokin** [comment]
Hi! The predicted maps look random, so looks like network is not training. Did you modify the optimizer properly? I suggest to replace current `Adam` to [`AdamW`](https://pytorch.org/docs/stable/generated/torch.optim.AdamW.html) and get rid of different learning rates for different layers, so you can just call: `optimizer = optim.AdamW(net.parameters())`.

**Comment by connernam** [comment]
As mentioned, the optimizer was changed from Adam to AdamW. However, iter is exceeding 40,000 but the loss does not decrease, and the ap is still zero. I am learning at 3080ti, and sometimes ap and ar are marked as 0.003 or similar. I've always used the pretrained model, so I can't determine if this situation is normal. 
Thank you always for the good advice :)

**Comment by connernam** [comment]
Loss is certainly gradually decreasing, and it is still AP=0.000. Sometimes the AP is not 0, and sometimes the AR is not 0.

Is that what AP looks like because learning is still in its early stages? Or I don't know if there is still a problem.

![image](https://user-images.githubusercontent.com/42915968/154217948-a81fd05f-f5e8-44bc-99ec-be7f2e154723.png)

**Comment by Daniil-Osokin** [comment]
That is strange. Can you share the full code (branch in a github repository) you are trying to run?

**Comment by connernam** [comment]
I made repo with full code [here](https://github.com/connernam/lightweight-human-pose-estimation.pytorch/tree/master). I accidentally pressed full request because I didn't use git. I'm sorry.

And it is an environment in which the code was executed.
Windows 10, Anaconda
![image](https://user-images.githubusercontent.com/42915968/154454344-149daf35-2429-4f8d-8d32-f1b876b85901.png)

**Comment by Daniil-Osokin** [comment]
The issue was that optimizer does not update the weights after each batch. Here is the fix: [fix_training.txt](https://github.com/Daniil-Osokin/lightweight-human-pose-estimation.pytorch/files/8094471/fix_training.txt). I have also changed input image normalization to fit MobileNetV3 and allowed pre-trained weights usage. You can apply the patch by `git apply fix_taining.txt` in the project root folder.

**Comment by KepingPan** [comment]
> [i've](https://github.com/connernam/lightweight-human-pose-estimation.pytorch/tree/master) made a repo here with the full code. i accidentally pressed the full request because i wasn't using git. i am sorry.
> 
> It is the environment in which the code is executed. Windows 10, Anaconda ![image](https://user-images.githubusercontent.com/42915968/154454344-149daf35-2429-4f8d-8d32-f1b876b85901.png)

Hey, I trained on your code, and it was 5% accurate.What is the accuracy rate of your training? I would be grateful if you could answer

**Comment by Daniil-Osokin** [comment]
I did not train this MobileNetV3 version, just checked that loss is reasonable and decreases over batches.

**Comment by connernam** [comment]
@KepingPan Currently, the learning progress is about 0.064 for the AP from iter=430,000. Although the results are not so good, it seems that tuning should be performed on the stride of the target image or on the cpm, init, refinement stage, etc.
Thank you @Daniil-Osokin for your help.

**Comment by KepingPan** [comment]
@connernam Thank you for your answer. I am also working on the improvement of feature network, so that we can communicate more.

**Comment by connernam** [comment]
@KepingPan If there is a better result, I would like you to share it. :)

**Comment by favor6** [comment]
Thank you very much for your help。
